File: sem2/po/PO-VirtualWorld-Python/simulator/organisms/animals/Antelope.py
from random import randint

from simulator.organisms.animals.Animal import Animal
from simulator.organisms.Organism import ORGANISM_TYPES

ANTELOPE_STRENGTH = 4
ANTELOPE_INITIATIVE = 4
ANTELOPE_SPEED = 2
ANTELOPE_ESCAPE_CHANCE_BASE = 10
ANTELOPE_ESCAPE_CHANCE_THRESHOLD = 5


class Antelope(Animal):
    def __init__(self, new_world, new_location):
        super().__init__(new_world, new_location, ANTELOPE_STRENGTH, ANTELOPE_INITIATIVE, ORGANISM_TYPES["Antelope"], ANTELOPE_SPEED)

    def spawn_offspring(self, new_location):
        return Antelope(self.world, new_location)

    def __str__(self):
        return "Antelope at " + str(self.location)

    def escape(self):
        escape_chance = randint(0, ANTELOPE_ESCAPE_CHANCE_BASE)
        if escape_chance > ANTELOPE_ESCAPE_CHANCE_THRESHOLD:
            new_location = self.get_map().get_first_free_neighbour(self.location)
            if new_location is not None:
                self.tlog(" escaped to " + str(new_location))
                self.update_location(new_location)
                return True
        else:
            return False

    def collision(self, invader):
        if invader.get_type() == self.get_type():
            self.mate()
        else:
            if not self.escape():
                self.fight(invader)
            else:
                invader.update_location(self.location)

File: sem2/po/po_proj2/src/studia/prog_ob/projekt2/world_simulator/organisms/animals/animals_subclasses/Antelope.java
package studia.prog_ob.projekt2.world_simulator.organisms.animals.animals_subclasses;


import studia.prog_ob.projekt2.world_simulator.RandomGenerator;
import studia.prog_ob.projekt2.world_simulator.SimLogger;
import studia.prog_ob.projekt2.world_simulator.World;
import studia.prog_ob.projekt2.world_simulator.maps.MapLocation;
import studia.prog_ob.projekt2.world_simulator.organisms.Organism;
import studia.prog_ob.projekt2.world_simulator.organisms.animals.Animal;

public class Antelope extends Animal {

    public static final int ANTELOPE_STRENGTH = 4;
    public static final int ANTELOPE_INITIATIVE = 4;
    public static final int ANTELOPE_SPEED = 2;
    public static final char ANTELOPE_SYMBOL = 'A';
    public static final int ANTELOPE_ESCAPE_CHANCE_BASE = 10;
    public static final int ANTELOPE_ESCEPE_THRESHOLD = 5;

    public Antelope(MapLocation new_map_location, World new_world) {
        super(ANTELOPE_STRENGTH, ANTELOPE_INITIATIVE, new_map_location, ANTELOPE_SYMBOL, new_world, ANTELOPE_SPEED);
    }

    @Override
    public Organism spawn_offspring(MapLocation location) {
        return getWorld().add_organism(new Antelope(location, getWorld()));
    }

    @Override
    public void collision(Organism invader) {
        if (invader instanceof Animal invading_animal) {
            if (same_type(invader)) {
                SimLogger.log(this + SimLogger.MATING_MESSAGE + " " + invader);
                mate();
                return;
            }

            if (!reflect_attack(invading_animal)) {
                MapLocation cl = new MapLocation(getMap_location());

                if (escape()) {
                    invading_animal.update_location(cl);
                } else {
                    fight(invading_animal);
                }
            }
        }
    }

    boolean escape() {
        int escape = RandomGenerator.rand_int(ANTELOPE_ESCAPE_CHANCE_BASE);

        if (escape >= ANTELOPE_ESCEPE_THRESHOLD) {
            for (int i = 1; i < getMap().DIRECTIONS_N + 1; i++) {
                MapLocation l = getMap().get_location_in_direction(getMap_location(), i, getSpeed());
                if (!getMap().out_of_map(l) && getMap().get_object(l) == null) {
                    SimLogger.log(this + SimLogger.FLEE_MESSAGE + " to " + l);
                    update_location(l);
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return "Antelope " + getMap_location() + " ";
    }
}
